<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Xiangyu Zhang SA20017912" />

<meta name="date" content="2020-12-17" />

<title>SimGenoData</title>

<style type="text/css">
a.anchor-section {margin-left: 10px; visibility: hidden; color: inherit;}
a.anchor-section::before {content: '#';}
.hasAnchor:hover a.anchor-section {visibility: visible;}
</style>
<script>// Anchor sections v1.0 written by Atsushi Yasumoto on Oct 3rd, 2020.
document.addEventListener('DOMContentLoaded', function() {
  // Do nothing if AnchorJS is used
  if (typeof window.anchors === 'object' && anchors.hasOwnProperty('hasAnchorJSLink')) {
    return;
  }

  const h = document.querySelectorAll('h1, h2, h3, h4, h5, h6');

  // Do nothing if sections are already anchored
  if (Array.from(h).some(x => x.classList.contains('hasAnchor'))) {
    return null;
  }

  // Use section id when pandoc runs with --section-divs
  const section_id = function(x) {
    return ((x.classList.contains('section') || (x.tagName === 'SECTION'))
            ? x.id : '');
  };

  // Add anchors
  h.forEach(function(x) {
    const id = x.id || section_id(x.parentElement);
    if (id === '') {
      return null;
    }
    let anchor = document.createElement('a');
    anchor.href = '#' + id;
    anchor.classList = ['anchor-section'];
    x.classList.add('hasAnchor');
    x.appendChild(anchor);
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SimGenoData</h1>
<h4 class="author">Xiangyu Zhang SA20017912</h4>
<h4 class="date">2020-12-17</h4>



<p>The following algorithms are based on the article: A Note on Generating Correlated Binary Variables (A. D. Lunn and S. J. Davies, Biometrika, Vol. 85, No. 2 (Jun., 1998), pp. 487-490). In this notes, A. D. Lunn and S. J. Davies concerns a simple, efficient generation of correlated binary data. The basic thought is summarized below:</p>
<p>Consider we have binary values <span class="math inline">\(X_{ij}\)</span> for <span class="math inline">\(i=1,...,k_{j}\)</span> and <span class="math inline">\(j=1,...,n\)</span>, where <span class="math inline">\(P(X_{ij})=1=p_{j}\)</span>. Let independent random variables <span class="math inline">\(Y_{ij}, Z_{j}\)</span> be distributed binomially <span class="math inline">\(Bi(1,p_{j})\)</span> and define <span class="math inline">\(X_{ij}=(1-U_{ij})Y_{ij}+U_{ij}Z_{j}\)</span>, where <span class="math inline">\(U_{ij}\)</span> are also independent and distributed binomiall <span class="math inline">\(Bi(1,r_{j})\)</span>. Then it is easy to show that the <span class="math inline">\(X_{ij}\)</span> are Bernoulli random variables with <span class="math inline">\(E(X_{ij})=p_{j}\)</span> and it follows that <span class="math inline">\(var(X_{ij})=p_{j}(1-p_{j})\)</span>. Now <span class="math display">\[E(X_{ij}X_{mj})=r^{2}_{j}p_{j}(1-p_{j})+p^{2}_{j}\quad (i \not= m), \quad  cov(X_{ij}, X_{mj})=r^{2}_{j}p_{j}(1-p_{j})\quad (i \not= m),\]</span> so that <span class="math inline">\(\rho_{j}=r^{2}_{j}\)</span>. This allows fast, efficient simulations to be carried out for a commonly occurring type of data. Each cluster of size <span class="math inline">\(k_{j}\)</span> requires the generation of only <span class="math inline">\(2k_{j}+1\)</span> independent Bernoulli variates.</p>
<div id="generating-correlated-binary-data" class="section level2">
<h2>Generating correlated binary data</h2>
<p>With this basic method, we can consider the following two algorithms which can generate one dimensional correlated binary data and high dimensional correlated binary data under certain probability of success, respectively.</p>
<p>One dimensional correlated binary data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">rcbern0&lt;-<span class="cf">function</span>(n, p, rho){</a>
<a class="sourceLine" id="cb1-2" title="2">    <span class="co">#Generate Correlated Bernoulli Distribution</span></a>
<a class="sourceLine" id="cb1-3" title="3">    <span class="co">#See &quot;A Note on Generating Correlated Binary Variables&quot;</span></a>
<a class="sourceLine" id="cb1-4" title="4">    <span class="co">#A. D. Lunn and S. J. Davies</span></a>
<a class="sourceLine" id="cb1-5" title="5">    <span class="co">#Biometrika</span></a>
<a class="sourceLine" id="cb1-6" title="6">    <span class="co">#Vol. 85, No. 2 (Jun., 1998), pp. 487-490 </span></a>
<a class="sourceLine" id="cb1-7" title="7">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.atomic</span>(p) <span class="op">||</span><span class="st"> </span><span class="kw">typeof</span>(p)<span class="op">!=</span><span class="st">&#39;double&#39;</span>) <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb1-8" title="8">    <span class="cf">if</span>(<span class="kw">length</span>(p)<span class="op">!=</span><span class="dv">1</span>) <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb1-9" title="9">    <span class="cf">if</span>((p<span class="op">&lt;</span><span class="dv">0</span>) <span class="op">|</span><span class="st"> </span>(p<span class="op">&gt;</span><span class="dv">1</span>)) <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb1-10" title="10">    </a>
<a class="sourceLine" id="cb1-11" title="11">    <span class="cf">if</span>((rho<span class="op">&lt;</span><span class="dv">0</span>) <span class="op">||</span><span class="st"> </span>(rho<span class="op">&gt;</span><span class="dv">1</span>)) <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb1-12" title="12">    Z&lt;-<span class="kw">rbinom</span>(<span class="dv">1</span>, <span class="dv">1</span>, p)</a>
<a class="sourceLine" id="cb1-13" title="13">    Y&lt;-<span class="kw">rbinom</span>(n, <span class="dv">1</span>, p)</a>
<a class="sourceLine" id="cb1-14" title="14">    U&lt;-<span class="kw">rbinom</span>(n, <span class="dv">1</span>, <span class="kw">sqrt</span>(rho))</a>
<a class="sourceLine" id="cb1-15" title="15">    X&lt;-(<span class="dv">1</span><span class="op">-</span>U)<span class="op">*</span>Y<span class="op">+</span>U<span class="op">*</span>Z</a>
<a class="sourceLine" id="cb1-16" title="16"></a>
<a class="sourceLine" id="cb1-17" title="17">    <span class="kw">return</span>(X)</a>
<a class="sourceLine" id="cb1-18" title="18">}</a></code></pre></div>
<p>High dimensional correlated binary data:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">rcbern&lt;-<span class="cf">function</span>(n, p, rho){</a>
<a class="sourceLine" id="cb2-2" title="2">    <span class="co">#Generate Correlated Bernoulli Distribution</span></a>
<a class="sourceLine" id="cb2-3" title="3">    <span class="co">#See &quot;A Note on Generating Correlated Binary Variables&quot;</span></a>
<a class="sourceLine" id="cb2-4" title="4">    <span class="co">#A. D. Lunn and S. J. Davies</span></a>
<a class="sourceLine" id="cb2-5" title="5">    <span class="co">#Biometrika</span></a>
<a class="sourceLine" id="cb2-6" title="6">    <span class="co">#Vol. 85, No. 2 (Jun., 1998), pp. 487-490 </span></a>
<a class="sourceLine" id="cb2-7" title="7">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.atomic</span>(p) <span class="op">||</span><span class="st"> </span><span class="kw">typeof</span>(p)<span class="op">!=</span><span class="st">&#39;double&#39;</span>) <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb2-8" title="8">    <span class="cf">if</span>(<span class="kw">sum</span>((p<span class="op">&lt;</span><span class="dv">0</span>) <span class="op">|</span><span class="st"> </span>(p<span class="op">&gt;</span><span class="dv">1</span>))<span class="op">!=</span><span class="dv">0</span>) <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb2-9" title="9"></a>
<a class="sourceLine" id="cb2-10" title="10">    m&lt;-<span class="kw">length</span>(p)</a>
<a class="sourceLine" id="cb2-11" title="11">    minP&lt;-<span class="kw">min</span>(p)</a>
<a class="sourceLine" id="cb2-12" title="12">    maxP&lt;-<span class="kw">max</span>(p)</a>
<a class="sourceLine" id="cb2-13" title="13">    rhoLimit&lt;-<span class="kw">sqrt</span>((minP<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>minP))<span class="op">/</span>(maxP<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>maxP)))</a>
<a class="sourceLine" id="cb2-14" title="14">    <span class="cf">if</span>((rho<span class="op">&lt;</span><span class="dv">0</span>) <span class="op">||</span><span class="st"> </span>(rho<span class="op">&gt;</span>rhoLimit)){</a>
<a class="sourceLine" id="cb2-15" title="15">        <span class="kw">cat</span>(<span class="kw">paste</span>(<span class="st">&#39;The range of rho is [&#39;</span>,<span class="dv">0</span>,<span class="st">&#39;,&#39;</span>,<span class="kw">round</span>(rhoLimit,<span class="dv">3</span>),<span class="st">&#39;]</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="dt">sep=</span><span class="st">&#39;&#39;</span>))</a>
<a class="sourceLine" id="cb2-16" title="16">        <span class="kw">cat</span>(<span class="st">&#39;rho is out-of-range</span><span class="ch">\n</span><span class="st">&#39;</span>)</a>
<a class="sourceLine" id="cb2-17" title="17">        <span class="kw">return</span>(<span class="ot">NaN</span>)</a>
<a class="sourceLine" id="cb2-18" title="18">    }</a>
<a class="sourceLine" id="cb2-19" title="19">    Pc&lt;-<span class="kw">sqrt</span>(minP<span class="op">*</span>maxP)<span class="op">/</span>(<span class="kw">sqrt</span>(minP<span class="op">*</span>maxP)<span class="op">+</span><span class="kw">sqrt</span>((<span class="dv">1</span><span class="op">-</span>minP)<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>maxP)))</a>
<a class="sourceLine" id="cb2-20" title="20">    Pa&lt;-<span class="kw">sqrt</span>(rho<span class="op">*</span>p<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">/</span>(Pc<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>Pc)))</a>
<a class="sourceLine" id="cb2-21" title="21">    <span class="cf">if</span>(<span class="kw">max</span>(Pa)<span class="op">==</span><span class="dv">1</span>){</a>
<a class="sourceLine" id="cb2-22" title="22">      X&lt;-<span class="kw">replicate</span>(n, <span class="kw">rep</span>(<span class="kw">rbinom</span>(<span class="dv">1</span>, <span class="dv">1</span>, p[<span class="dv">1</span>]), m))</a>
<a class="sourceLine" id="cb2-23" title="23">    }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb2-24" title="24">    Pb&lt;-(p<span class="op">-</span>Pa<span class="op">*</span>Pc)<span class="op">/</span>(<span class="dv">1</span><span class="op">-</span>Pa)</a>
<a class="sourceLine" id="cb2-25" title="25">    X&lt;-<span class="kw">replicate</span>(n, {</a>
<a class="sourceLine" id="cb2-26" title="26">        U&lt;-<span class="kw">rbinom</span>(m, <span class="dv">1</span>, Pa)</a>
<a class="sourceLine" id="cb2-27" title="27">        Y&lt;-<span class="kw">rbinom</span>(m, <span class="dv">1</span>, Pb)</a>
<a class="sourceLine" id="cb2-28" title="28">        Z&lt;-<span class="kw">rbinom</span>(<span class="dv">1</span>, <span class="dv">1</span>, Pc)</a>
<a class="sourceLine" id="cb2-29" title="29">        (<span class="dv">1</span><span class="op">-</span>U)<span class="op">*</span>Y<span class="op">+</span>U<span class="op">*</span>Z</a>
<a class="sourceLine" id="cb2-30" title="30">    })</a>
<a class="sourceLine" id="cb2-31" title="31">    }</a>
<a class="sourceLine" id="cb2-32" title="32">    X&lt;-<span class="kw">t</span>(X)</a>
<a class="sourceLine" id="cb2-33" title="33"></a>
<a class="sourceLine" id="cb2-34" title="34">    <span class="kw">return</span>(X)</a>
<a class="sourceLine" id="cb2-35" title="35">}</a></code></pre></div>
<p>We also provide some simple examples for these algorithms:</p>
</div>
<div id="example-1" class="section level2">
<h2>Example 1</h2>
<p>If we want to generate 10 binary samples with correlationship 0.2 and probability of success 0.5, we can use:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw">rcbern0</span>(<span class="dv">10</span>,<span class="fl">0.5</span>,<span class="fl">0.2</span>)</a></code></pre></div>
<pre><code>##  [1] 0 0 0 0 0 0 0 1 0 0</code></pre>
<p>which returns a vector, or</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw">rcbern</span>(<span class="dv">10</span>,<span class="fl">0.5</span>,<span class="fl">0.2</span>)</a></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    0    0    0    0    1    1    1    0    1     1</code></pre>
<p>which returns a matrix.</p>
</div>
<div id="example-2" class="section level2">
<h2>Example 2</h2>
<p>If we want to generate 10 3-dim binary samples with correlationship 0.2 and probability of success p=c(0.1,0.2,0.3), we can use:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">p&lt;-<span class="kw">c</span>(<span class="fl">0.1</span>,<span class="fl">0.2</span>,<span class="fl">0.3</span>)</a>
<a class="sourceLine" id="cb7-2" title="2"><span class="kw">rcbern</span>(<span class="dv">10</span>,p,<span class="fl">0.2</span>)</a></code></pre></div>
<pre><code>##       [,1] [,2] [,3]
##  [1,]    0    0    0
##  [2,]    0    0    1
##  [3,]    0    0    0
##  [4,]    0    0    0
##  [5,]    0    1    1
##  [6,]    0    0    0
##  [7,]    0    0    0
##  [8,]    0    0    0
##  [9,]    0    1    0
## [10,]    0    1    1</code></pre>
<p>which returns a 10*3 matrix.</p>
</div>
<div id="genotype-data" class="section level2">
<h2>Genotype data</h2>
<p>If we specify the probabilities of success as the minor allele frequencies (MAFs) and let the variable <span class="math inline">\(X\)</span> be the sum of two independent binary variables. We obtain the following algorithm which generate the genotypes of <span class="math inline">\(n\)</span> individuals with pairwise correlationship <span class="math inline">\(\rho\)</span> and MAFs <span class="math inline">\(f\)</span>. The number of SNPs can be specified by the length of input <span class="math inline">\(f\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">simGenoData&lt;-<span class="cf">function</span>(n, f, rho){</a>
<a class="sourceLine" id="cb9-2" title="2">    m&lt;-<span class="kw">length</span>(f)</a>
<a class="sourceLine" id="cb9-3" title="3">    X&lt;-<span class="kw">matrix</span>(<span class="dv">0</span>, n, m)</a>
<a class="sourceLine" id="cb9-4" title="4">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m){</a>
<a class="sourceLine" id="cb9-5" title="5">        <span class="cf">if</span>(rho<span class="op">!=</span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb9-6" title="6">            <span class="co">#X[,i]&lt;-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))       </span></a>
<a class="sourceLine" id="cb9-7" title="7">            X[,i]&lt;-<span class="kw">rcbern0</span>(n, f[i], rho)<span class="op">+</span><span class="kw">rcbern0</span>(n, f[i], rho)</a>
<a class="sourceLine" id="cb9-8" title="8">        }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb9-9" title="9">            X[,i]&lt;-<span class="kw">rbinom</span>(n,<span class="dv">2</span>,f[i])</a>
<a class="sourceLine" id="cb9-10" title="10">        }</a>
<a class="sourceLine" id="cb9-11" title="11">    }</a>
<a class="sourceLine" id="cb9-12" title="12">    <span class="kw">return</span>(X)</a>
<a class="sourceLine" id="cb9-13" title="13">}</a></code></pre></div>
</div>
<div id="example-3" class="section level2">
<h2>Example 3</h2>
<p>If we want to generate the genotypes of <span class="math inline">\(n=2\)</span> individuals with pairwise correlationship <span class="math inline">\(\rho=0.1\)</span> at 10 SNPs with MAFs 0.2, we can:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">f&lt;-<span class="kw">rep</span>(<span class="fl">0.2</span>,<span class="dv">10</span>)</a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">simGenoData</span>(<span class="dv">2</span>,f,<span class="fl">0.1</span>)</a></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    0    2    0    1    1    1    1    0    0     0
## [2,]    0    2    1    0    0    1    2    0    0     1</code></pre>
<p>which returns a 2*10 matrix with each line represent the genotype for one individual.</p>
</div>
<div id="gwas-simulation" class="section level2">
<h2>Gwas Simulation</h2>
<p>We also applies these algorithms in Genome-Wide Association Studies (GWAS) simulations. We consider the following three conditions. The first one is that we want to generate the genotypes when the pairwise correlationship is <span class="math inline">\(\rho\)</span>, regardless of the individuals are select from cases or controls. The algorithm below is designed for this condition with input <span class="math inline">\(n0,n1,f0,f1,rho\)</span> be the case size, control size, MAFs of cases, MAFs of controls, pairwise correlationship, respectively. The output includes the the genotype matrix and label of cases or controls (the same for the other two conditions).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">simGwasData&lt;-<span class="cf">function</span>(n0, n1, f0, f1, rho){</a>
<a class="sourceLine" id="cb12-2" title="2">  n&lt;-n0<span class="op">+</span>n1</a>
<a class="sourceLine" id="cb12-3" title="3">  m&lt;-<span class="kw">length</span>(f0)</a>
<a class="sourceLine" id="cb12-4" title="4">  X&lt;-<span class="kw">matrix</span>(<span class="dv">0</span>, n, m)</a>
<a class="sourceLine" id="cb12-5" title="5">  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m){</a>
<a class="sourceLine" id="cb12-6" title="6">    <span class="cf">if</span>(rho<span class="op">!=</span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb12-7" title="7">      <span class="co">#X[,i]&lt;-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))     </span></a>
<a class="sourceLine" id="cb12-8" title="8">      f&lt;-<span class="kw">c</span>(<span class="kw">rep</span>(f0[i],n0), <span class="kw">rep</span>(f1[i], n1))</a>
<a class="sourceLine" id="cb12-9" title="9">      X[,i]&lt;-<span class="kw">rcbern</span>(<span class="dv">1</span>, f, rho)<span class="op">+</span><span class="kw">rcbern</span>(<span class="dv">1</span>, f, rho)  </a>
<a class="sourceLine" id="cb12-10" title="10">    }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb12-11" title="11">      X[,i]&lt;-<span class="kw">c</span>(<span class="kw">rbinom</span>(n0, <span class="dv">2</span>, f0[i]), <span class="kw">rbinom</span>(n1, <span class="dv">2</span>, f1[i]))</a>
<a class="sourceLine" id="cb12-12" title="12">    }</a>
<a class="sourceLine" id="cb12-13" title="13">  }</a>
<a class="sourceLine" id="cb12-14" title="14">  y&lt;-<span class="kw">c</span>(<span class="kw">rep</span>(<span class="dv">0</span>, n0), <span class="kw">rep</span>(<span class="dv">1</span>, n1))</a>
<a class="sourceLine" id="cb12-15" title="15">  <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">X=</span>X, <span class="dt">y=</span>y))</a>
<a class="sourceLine" id="cb12-16" title="16">}</a></code></pre></div>
<p>The second condition is that we want to generate the genotypes when the pairwise correlationship is <span class="math inline">\(\rho\)</span> within cases and controls, while there is no correlations between a case and a control. The algorithm below is designed for this condition with input <span class="math inline">\(n0,n1,f0,f1,rho\)</span> be the case size, control size, MAFs of cases, MAFs of controls, pairwise correlationship, respectively.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">simGwasDataI&lt;-<span class="cf">function</span>(n0, n1, f0, f1, rho){</a>
<a class="sourceLine" id="cb13-2" title="2">  n&lt;-n0<span class="op">+</span>n1</a>
<a class="sourceLine" id="cb13-3" title="3">  m&lt;-<span class="kw">length</span>(f0)</a>
<a class="sourceLine" id="cb13-4" title="4">  X&lt;-<span class="kw">matrix</span>(<span class="dv">0</span>, n, m)</a>
<a class="sourceLine" id="cb13-5" title="5">  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m){</a>
<a class="sourceLine" id="cb13-6" title="6">    <span class="cf">if</span>(rho<span class="op">!=</span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb13-7" title="7">      <span class="co">#X[,i]&lt;-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))</span></a>
<a class="sourceLine" id="cb13-8" title="8">      f0i&lt;-<span class="kw">rep</span>(f0[i], n0)</a>
<a class="sourceLine" id="cb13-9" title="9">      f1i&lt;-<span class="kw">rep</span>(f1[i], n1)</a>
<a class="sourceLine" id="cb13-10" title="10">      X[,i]&lt;-<span class="kw">c</span>(<span class="kw">rcbern</span>(<span class="dv">1</span>, f0i, rho)<span class="op">+</span><span class="kw">rcbern</span>(<span class="dv">1</span>, f0i, rho), <span class="kw">rcbern</span>(<span class="dv">1</span>, f1i, rho)<span class="op">+</span><span class="kw">rcbern</span>(<span class="dv">1</span>, f1i, rho))</a>
<a class="sourceLine" id="cb13-11" title="11">    }<span class="cf">else</span>{</a>
<a class="sourceLine" id="cb13-12" title="12">      X[,i]&lt;-<span class="kw">c</span>(<span class="kw">rbinom</span>(n0, <span class="dv">2</span>, f0[i]), <span class="kw">rbinom</span>(n1, <span class="dv">2</span>, f1[i]))</a>
<a class="sourceLine" id="cb13-13" title="13">    }</a>
<a class="sourceLine" id="cb13-14" title="14">  }</a>
<a class="sourceLine" id="cb13-15" title="15">  y&lt;-<span class="kw">c</span>(<span class="kw">rep</span>(<span class="dv">0</span>, n0), <span class="kw">rep</span>(<span class="dv">1</span>, n1))</a>
<a class="sourceLine" id="cb13-16" title="16">  <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">X=</span>X, <span class="dt">y=</span>y))</a>
<a class="sourceLine" id="cb13-17" title="17">}</a></code></pre></div>
<p>The third condition is that we want to generate the genotypes when the pairwise correlationship is <span class="math inline">\(\rho0\)</span> within cases and <span class="math inline">\(\rho1\)</span> within controls, while there is no correlations between a case and a control. The algorithm below is designed for this condition with input <span class="math inline">\(n0,n1,f0,f1,rho0,rho1\)</span> be the case size, control size, MAFs of cases, MAFs of controls, pairwise correlationship between cases, pairwise correlationship between controls, respectively.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">simGwasDataII&lt;-<span class="cf">function</span>(n0, n1, f0, f1, rho0, rho1){</a>
<a class="sourceLine" id="cb14-2" title="2">  n&lt;-n0<span class="op">+</span>n1</a>
<a class="sourceLine" id="cb14-3" title="3">  m&lt;-<span class="kw">length</span>(f0)</a>
<a class="sourceLine" id="cb14-4" title="4">  </a>
<a class="sourceLine" id="cb14-5" title="5">  X&lt;-<span class="kw">matrix</span>(<span class="dv">0</span>, n, m)</a>
<a class="sourceLine" id="cb14-6" title="6">  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m){</a>
<a class="sourceLine" id="cb14-7" title="7">    <span class="cf">if</span>(rho0<span class="op">!=</span><span class="dv">0</span> <span class="op">||</span><span class="st"> </span>rho1<span class="op">!=</span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb14-8" title="8">      <span class="co">#X[,i]&lt;-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))</span></a>
<a class="sourceLine" id="cb14-9" title="9">      f0i&lt;-<span class="kw">rep</span>(f0[i], n0)</a>
<a class="sourceLine" id="cb14-10" title="10">      f1i&lt;-<span class="kw">rep</span>(f1[i], n1)</a>
<a class="sourceLine" id="cb14-11" title="11">      X[,i]&lt;-<span class="kw">c</span>(<span class="kw">rcbern</span>(<span class="dv">1</span>, f0i, rho0)<span class="op">+</span><span class="kw">rcbern</span>(<span class="dv">1</span>, f0i, rho0), <span class="kw">rcbern</span>(<span class="dv">1</span>, f1i, rho1)<span class="op">+</span><span class="kw">rcbern</span>(<span class="dv">1</span>, f1i, rho1))</a>
<a class="sourceLine" id="cb14-12" title="12">    }</a>
<a class="sourceLine" id="cb14-13" title="13">    <span class="cf">else</span>{</a>
<a class="sourceLine" id="cb14-14" title="14">      X[,i]&lt;-<span class="kw">c</span>(<span class="kw">rbinom</span>(n0, <span class="dv">2</span>, f0[i]), <span class="kw">rbinom</span>(n1, <span class="dv">2</span>, f1[i]))</a>
<a class="sourceLine" id="cb14-15" title="15">    }</a>
<a class="sourceLine" id="cb14-16" title="16">  }</a>
<a class="sourceLine" id="cb14-17" title="17">  y&lt;-<span class="kw">c</span>(<span class="kw">rep</span>(<span class="dv">0</span>, n0), <span class="kw">rep</span>(<span class="dv">1</span>, n1))</a>
<a class="sourceLine" id="cb14-18" title="18">  <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">X=</span>X, <span class="dt">y=</span>y))</a>
<a class="sourceLine" id="cb14-19" title="19">  <span class="co"># return(X)</span></a>
<a class="sourceLine" id="cb14-20" title="20">}</a></code></pre></div>
<p>We provide one example for each of these conditions:</p>
</div>
<div id="example-4-condition-1" class="section level2">
<h2>Example 4: Condition 1</h2>
<p>If we want to generate the genotypes of 10 cases and 10 controls with pairwise correlationship <span class="math inline">\(\rho =0.1\)</span> for all individuals at 20 SNPs with MAFs generated uniformly from [0.05,0.5], we can:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">f0&lt;-f1&lt;-<span class="kw">runif</span>(<span class="dv">20</span>,<span class="fl">0.05</span>,<span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb15-2" title="2"><span class="kw">simGwasData</span>(<span class="dv">10</span>,<span class="dv">10</span>,f0,f1,<span class="fl">0.1</span>)</a></code></pre></div>
<pre><code>## $X
##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
##  [1,]    1    0    1    1    0    0    1    1    0     1     0     1     1
##  [2,]    1    1    0    0    0    2    0    0    0     0     0     2     1
##  [3,]    0    1    1    1    0    0    2    1    0     2     0     1     1
##  [4,]    1    0    0    1    0    0    1    0    0     1     0     2     1
##  [5,]    1    0    1    1    0    0    2    0    0     1     0     1     1
##  [6,]    1    1    1    1    0    0    0    0    0     0     0     1     0
##  [7,]    0    0    2    0    0    0    1    1    1     1     0     1     0
##  [8,]    0    0    0    1    0    0    2    0    0     2     0     1     0
##  [9,]    2    0    0    1    0    0    2    0    0     0     0     2     1
## [10,]    1    1    2    0    0    0    2    0    0     1     0     1     1
## [11,]    2    0    1    1    0    0    2    0    0     0     0     1     2
## [12,]    2    0    1    0    1    0    2    1    0     1     0     2     1
## [13,]    0    0    1    0    0    0    2    0    0     0     0     2     1
## [14,]    1    1    1    0    0    1    1    0    0     1     0     1     1
## [15,]    1    1    0    1    0    0    1    0    0     1     0     1     1
## [16,]    2    0    1    0    0    0    2    1    0     2     1     2     0
## [17,]    1    0    1    1    0    2    1    0    1     0     0     1     1
## [18,]    1    0    0    1    0    0    1    0    0     1     0     1     1
## [19,]    1    0    0    0    0    0    2    1    0     1     0     0     0
## [20,]    1    0    2    1    0    0    2    0    0     0     0     1     0
##       [,14] [,15] [,16] [,17] [,18] [,19] [,20]
##  [1,]     1     1     1     1     0     1     0
##  [2,]     2     1     0     1     0     1     0
##  [3,]     2     0     0     0     1     0     1
##  [4,]     2     0     2     0     0     0     1
##  [5,]     0     0     0     0     0     0     0
##  [6,]     2     1     1     1     0     1     0
##  [7,]     1     1     0     0     0     0     1
##  [8,]     1     2     1     0     0     1     1
##  [9,]     2     0     0     1     1     1     0
## [10,]     1     0     1     0     0     1     0
## [11,]     2     2     0     1     1     1     1
## [12,]     1     2     1     1     1     2     0
## [13,]     2     1     1     0     0     1     1
## [14,]     1     0     0     0     0     1     0
## [15,]     1     1     0     0     1     1     0
## [16,]     0     1     0     1     0     0     0
## [17,]     1     2     0     1     0     1     1
## [18,]     1     1     0     0     0     1     0
## [19,]     1     1     1     0     1     0     2
## [20,]     1     0     0     0     0     1     0
## 
## $y
##  [1] 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1</code></pre>
<p>If we want to generate the genotypes of 10 cases and 10 controls with pairwise correlationship <span class="math inline">\(\rho =0.1\)</span> within cases and controls at 20 SNPs with MAFs generated uniformly from [0.05,0.5], we can:</p>
</div>
<div id="example-5-condition-2" class="section level2">
<h2>Example 5: Condition 2</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">f0&lt;-f1&lt;-<span class="kw">runif</span>(<span class="dv">20</span>,<span class="fl">0.05</span>,<span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb17-2" title="2"><span class="kw">simGwasDataI</span>(<span class="dv">10</span>,<span class="dv">10</span>,f0,f1,<span class="fl">0.1</span>)</a></code></pre></div>
<pre><code>## $X
##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
##  [1,]    0    0    1    0    0    0    0    1    0     1     2     0     0
##  [2,]    2    0    0    1    0    0    0    1    0     1     1     0     1
##  [3,]    1    0    0    0    0    0    0    1    0     0     1     0     0
##  [4,]    0    0    1    0    0    0    0    2    0     0     1     1     1
##  [5,]    1    0    0    1    0    1    0    1    0     0     1     0     2
##  [6,]    1    0    0    0    0    0    0    1    0     0     0     1     2
##  [7,]    0    0    0    0    0    0    0    2    1     1     1     0     1
##  [8,]    1    0    1    1    0    0    0    1    0     0     1     1     0
##  [9,]    1    0    1    0    0    0    0    0    0     0     1     0     2
## [10,]    2    1    0    0    0    0    0    1    0     0     2     0     1
## [11,]    0    1    0    0    0    0    0    0    0     0     1     1     1
## [12,]    0    1    0    0    0    0    0    0    0     0     1     1     0
## [13,]    0    1    0    0    0    0    1    2    1     2     0     1     0
## [14,]    0    0    0    0    0    0    0    0    0     1     0     1     1
## [15,]    1    0    0    0    0    0    0    2    0     0     1     1     1
## [16,]    0    0    1    1    0    0    0    1    0     0     1     0     1
## [17,]    0    0    0    1    0    1    0    0    0     2     1     1     1
## [18,]    0    1    1    0    0    0    1    0    0     1     1     0     1
## [19,]    0    0    0    0    0    0    1    0    0     2     0     0     0
## [20,]    1    0    1    0    0    0    0    1    0     0     0     0     0
##       [,14] [,15] [,16] [,17] [,18] [,19] [,20]
##  [1,]     1     0     1     0     1     1     0
##  [2,]     0     0     2     0     0     0     0
##  [3,]     0     1     2     0     1     0     1
##  [4,]     1     0     1     0     1     1     1
##  [5,]     1     1     0     0     0     0     1
##  [6,]     2     0     0     0     1     1     0
##  [7,]     0     0     1     0     2     1     1
##  [8,]     1     0     1     0     1     1     1
##  [9,]     1     0     0     1     0     1     1
## [10,]     1     1     1     0     1     1     0
## [11,]     0     1     0     0     1     0     0
## [12,]     0     1     0     0     0     1     0
## [13,]     1     1     0     0     1     1     0
## [14,]     1     1     0     1     1     0     0
## [15,]     0     0     1     1     1     1     0
## [16,]     2     0     0     1     1     1     1
## [17,]     1     0     1     0     2     0     0
## [18,]     1     1     0     0     1     1     0
## [19,]     0     1     0     1     2     0     0
## [20,]     1     1     0     0     0     0     1
## 
## $y
##  [1] 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1</code></pre>
<p>If we want to generate the genotypes of 10 cases and 10 controls with pairwise correlationship 0.1 within cases and 0.2 within controls at 20 SNPs with MAFs generated uniformly from [0.05,0.5], we can:</p>
</div>
<div id="example-6-condition-3" class="section level2">
<h2>Example 6: Condition 3</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">f0=f1&lt;-<span class="kw">runif</span>(<span class="dv">20</span>,<span class="fl">0.05</span>,<span class="fl">0.5</span>)</a>
<a class="sourceLine" id="cb19-2" title="2"><span class="kw">simGwasDataII</span>(<span class="dv">10</span>,<span class="dv">10</span>,f0,f1,<span class="fl">0.1</span>,<span class="fl">0.2</span>)</a></code></pre></div>
<pre><code>## $X
##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
##  [1,]    1    0    0    0    1    1    0    0    1     0     2     1     1
##  [2,]    0    1    0    0    0    2    0    1    1     0     0     0     1
##  [3,]    1    0    1    0    0    1    0    1    0     1     1     2     1
##  [4,]    0    0    0    0    0    1    1    0    1     0     0     1     1
##  [5,]    0    1    1    0    0    0    0    1    0     0     1     1     0
##  [6,]    1    1    0    1    0    2    1    0    0     0     1     2     0
##  [7,]    1    0    1    1    0    1    0    0    0     0     1     0     1
##  [8,]    2    1    0    0    1    0    0    1    0     0     1     1     1
##  [9,]    0    0    0    1    0    0    0    0    0     0     0     1     2
## [10,]    0    0    1    0    0    1    0    0    0     0     1     1     2
## [11,]    1    2    0    1    0    2    2    1    1     0     2     0     0
## [12,]    0    1    0    1    0    0    0    0    1     0     2     0     0
## [13,]    0    0    0    1    0    0    1    0    1     0     2     0     2
## [14,]    1    1    0    1    0    2    1    1    1     0     2     0     1
## [15,]    0    2    0    0    0    0    1    0    1     0     0     0     0
## [16,]    0    0    0    0    1    0    1    0    1     1     2     1     0
## [17,]    0    1    0    2    0    0    1    0    1     0     1     0     0
## [18,]    0    1    0    1    1    1    1    0    1     0     2     0     2
## [19,]    0    2    0    1    0    1    1    0    0     0     1     2     0
## [20,]    1    2    0    2    0    1    1    0    1     0     1     0     0
##       [,14] [,15] [,16] [,17] [,18] [,19] [,20]
##  [1,]     0     0     0     0     1     0     1
##  [2,]     1     1     1     0     1     2     0
##  [3,]     1     0     0     1     1     1     1
##  [4,]     1     1     0     1     0     1     2
##  [5,]     0     2     1     0     1     0     2
##  [6,]     1     1     1     0     0     1     1
##  [7,]     0     0     0     0     0     1     0
##  [8,]     1     0     1     0     1     0     1
##  [9,]     0     1     0     0     1     0     1
## [10,]     0     0     0     0     0     0     1
## [11,]     1     1     0     0     1     1     1
## [12,]     1     1     0     0     0     1     1
## [13,]     1     1     0     1     1     0     1
## [14,]     2     1     0     0     0     0     2
## [15,]     0     1     2     0     0     2     0
## [16,]     2     1     0     1     0     0     1
## [17,]     0     2     0     0     0     1     1
## [18,]     0     0     0     0     0     1     0
## [19,]     0     1     0     0     1     1     1
## [20,]     0     2     1     0     0     0     2
## 
## $y
##  [1] 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
