---
title: "SimGenoData"
author: "Xiangyu Zhang SA20017912"
header-includes:
   - \usepackage{bbm}
   - \usepackage[utf8]{inputenc}
   - \usepackage{amsthm}
   - \usepackage{amssymb}
   - \usepackage{mathtools}
   - \usepackage{bbm}
   - \usepackage{bm}
   - \usepackage{mathrsfs}
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SimGenoData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


The following algorithms are based on the article: A Note on Generating Correlated Binary Variables (A. D. Lunn and S. J. Davies, Biometrika, Vol. 85, No. 2 (Jun., 1998), pp. 487-490). In this notes,  A. D. Lunn and S. J. Davies concerns a simple, efficient generation of correlated binary data. The basic thought is summarized below:

Consider we have binary values $X_{ij}$ for $i=1,...,k_{j}$ and $j=1,...,n$, where $P(X_{ij})=1=p_{j}$. Let independent random variables $Y_{ij}, Z_{j}$ be distributed binomially $Bi(1,p_{j})$ and define $X_{ij}=(1-U_{ij})Y_{ij}+U_{ij}Z_{j}$, where $U_{ij}$ are also independent and distributed binomiall $Bi(1,r_{j})$. Then it is easy to show that the $X_{ij}$ are Bernoulli random variables with $E(X_{ij})=p_{j}$ and it follows that $var(X_{ij})=p_{j}(1-p_{j})$. Now
$$E(X_{ij}X_{mj})=r^{2}_{j}p_{j}(1-p_{j})+p^{2}_{j}\quad (i \not= m), \quad  cov(X_{ij}, X_{mj})=r^{2}_{j}p_{j}(1-p_{j})\quad (i \not= m),$$
so that $\rho_{j}=r^{2}_{j}$. This allows fast, efficient simulations to be carried out for a commonly occurring type of data. Each cluster of size $k_{j}$ requires the generation of only $2k_{j}+1$ independent Bernoulli variates.

## Generating correlated binary data

With this basic method, we can consider the following two algorithms which can generate one dimensional correlated binary data and high dimensional correlated binary data under certain probability of success, respectively.

One dimensional correlated binary data:
```{r}
rcbern0<-function(n, p, rho){
	#Generate Correlated Bernoulli Distribution
	#See "A Note on Generating Correlated Binary Variables"
	#A. D. Lunn and S. J. Davies
	#Biometrika
	#Vol. 85, No. 2 (Jun., 1998), pp. 487-490 
	if(!is.atomic(p) || typeof(p)!='double') return(NaN)
	if(length(p)!=1) return(NaN)
	if((p<0) | (p>1)) return(NaN)
	
	if((rho<0) || (rho>1)) return(NaN)
	Z<-rbinom(1, 1, p)
	Y<-rbinom(n, 1, p)
	U<-rbinom(n, 1, sqrt(rho))
	X<-(1-U)*Y+U*Z

	return(X)
}
```

High dimensional correlated binary data:
```{r}
rcbern<-function(n, p, rho){
	#Generate Correlated Bernoulli Distribution
	#See "A Note on Generating Correlated Binary Variables"
	#A. D. Lunn and S. J. Davies
	#Biometrika
	#Vol. 85, No. 2 (Jun., 1998), pp. 487-490 
	if(!is.atomic(p) || typeof(p)!='double') return(NaN)
	if(sum((p<0) | (p>1))!=0) return(NaN)

	m<-length(p)
	minP<-min(p)
	maxP<-max(p)
	rhoLimit<-sqrt((minP/(1-minP))/(maxP/(1-maxP)))
	if((rho<0) || (rho>rhoLimit)){
		cat(paste('The range of rho is [',0,',',round(rhoLimit,3),']\n', sep=''))
		cat('rho is out-of-range\n')
		return(NaN)
	}
	Pc<-sqrt(minP*maxP)/(sqrt(minP*maxP)+sqrt((1-minP)*(1-maxP)))
	Pa<-sqrt(rho*p*(1-p)/(Pc*(1-Pc)))
	if(max(Pa)==1){
	  X<-replicate(n, rep(rbinom(1, 1, p[1]), m))
	}else{
	Pb<-(p-Pa*Pc)/(1-Pa)
	X<-replicate(n, {
		U<-rbinom(m, 1, Pa)
		Y<-rbinom(m, 1, Pb)
		Z<-rbinom(1, 1, Pc)
		(1-U)*Y+U*Z
	})
	}
	X<-t(X)

	return(X)
}
```

We also provide some simple examples for these algorithms:

## Example 1
If we want to generate 10 binary samples with correlationship 0.2 and probability of success 0.5, we can use:
```{r}
rcbern0(10,0.5,0.2)
```

which returns a vector, or
```{r}
rcbern(10,0.5,0.2)
```

which returns a matrix.

## Example 2
If we want to generate 10 3-dim binary samples with correlationship 0.2 and probability of success p=c(0.1,0.2,0.3), we can use:
```{r}
p<-c(0.1,0.2,0.3)
rcbern(10,p,0.2)
```

which returns a 10*3 matrix.

## Genotype data

If we specify the probabilities of success as the minor allele frequencies (MAFs) and let the variable $X$ be the sum of two independent binary variables. We obtain the following algorithm which generate the genotypes of $n$ individuals with pairwise correlationship $\rho$ and MAFs $f$. The number of SNPs can be specified by the length of input $f$.

```{r}
simGenoData<-function(n, f, rho){
	m<-length(f)
	X<-matrix(0, n, m)
	for(i in 1:m){
		if(rho!=0){
			#X[,i]<-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))		
			X[,i]<-rcbern0(n, f[i], rho)+rcbern0(n, f[i], rho)
		}else{
			X[,i]<-rbinom(n,2,f[i])
		}
	}
	return(X)
}
```

## Example 3

If we want to generate the genotypes of $n=2$ individuals with pairwise correlationship $\rho=0.1$ at 10 SNPs with MAFs 0.2, we can:

```{r}
f<-rep(0.2,10)
simGenoData(2,f,0.1)
```

which returns a 2*10 matrix with each line represent the genotype for one individual.


## Gwas Simulation

We also applies these algorithms in Genome-Wide Association Studies (GWAS) simulations. We consider the following three conditions. The first one is that we want to generate the genotypes when the pairwise correlationship is $\rho$, regardless of the individuals are select from cases or controls. The algorithm below is designed for this condition with input $n0,n1,f0,f1,rho$ be the  case size, control size, MAFs of cases, MAFs of controls, pairwise correlationship, respectively. The output includes the the genotype matrix and label of cases or controls (the same for the other two conditions). 

```{r}
simGwasData<-function(n0, n1, f0, f1, rho){
  n<-n0+n1
  m<-length(f0)
  X<-matrix(0, n, m)
  for(i in 1:m){
    if(rho!=0){
      #X[,i]<-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))		
      f<-c(rep(f0[i],n0), rep(f1[i], n1))
      X[,i]<-rcbern(1, f, rho)+rcbern(1, f, rho)  
    }else{
      X[,i]<-c(rbinom(n0, 2, f0[i]), rbinom(n1, 2, f1[i]))
    }
  }
  y<-c(rep(0, n0), rep(1, n1))
  return(list(X=X, y=y))
}
```

The second condition is that we want to generate the genotypes when the pairwise correlationship is $\rho$ within cases and controls, while there is no correlations between a case and a control. The algorithm below is designed for this condition with input $n0,n1,f0,f1,rho$ be the  case size, control size, MAFs of cases, MAFs of controls, pairwise correlationship, respectively. 

```{r}
simGwasDataI<-function(n0, n1, f0, f1, rho){
  n<-n0+n1
  m<-length(f0)
  X<-matrix(0, n, m)
  for(i in 1:m){
    if(rho!=0){
      #X[,i]<-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))
      f0i<-rep(f0[i], n0)
      f1i<-rep(f1[i], n1)
      X[,i]<-c(rcbern(1, f0i, rho)+rcbern(1, f0i, rho), rcbern(1, f1i, rho)+rcbern(1, f1i, rho))
    }else{
      X[,i]<-c(rbinom(n0, 2, f0[i]), rbinom(n1, 2, f1[i]))
    }
  }
  y<-c(rep(0, n0), rep(1, n1))
  return(list(X=X, y=y))
}
```


The third condition is that we want to generate the genotypes when the pairwise correlationship is $\rho0$ within cases and $\rho1$ within controls, while there is no correlations between a case and a control. The algorithm below is designed for this condition with input $n0,n1,f0,f1,rho0,rho1$ be the  case size, control size, MAFs of cases, MAFs of controls, pairwise correlationship between cases, pairwise correlationship between controls, respectively. 

```{r}
simGwasDataII<-function(n0, n1, f0, f1, rho0, rho1){
  n<-n0+n1
  m<-length(f0)
  
  X<-matrix(0, n, m)
  for(i in 1:m){
    if(rho0!=0 || rho1!=0){
      #X[,i]<-colSums(rmvbin(2, rep(f[i], n), bincorr=(1-rho)*diag(n)+rho))
      f0i<-rep(f0[i], n0)
      f1i<-rep(f1[i], n1)
      X[,i]<-c(rcbern(1, f0i, rho0)+rcbern(1, f0i, rho0), rcbern(1, f1i, rho1)+rcbern(1, f1i, rho1))
    }
    else{
      X[,i]<-c(rbinom(n0, 2, f0[i]), rbinom(n1, 2, f1[i]))
    }
  }
  y<-c(rep(0, n0), rep(1, n1))
  return(list(X=X, y=y))
  # return(X)
}
```

We provide one example for each of these conditions:

## Example 4: Condition 1

If we want to generate the genotypes of 10 cases and 10 controls with pairwise correlationship $\rho =0.1$ for all individuals at 20 SNPs with MAFs generated uniformly from [0.05,0.5], we can:

```{r}
f0<-f1<-runif(20,0.05,0.5)
simGwasData(10,10,f0,f1,0.1)
```

If we want to generate the genotypes of 10 cases and 10 controls with pairwise correlationship $\rho =0.1$ within cases and controls at 20 SNPs with MAFs generated uniformly from [0.05,0.5], we can:

## Example 5: Condition 2

```{r}
f0<-f1<-runif(20,0.05,0.5)
simGwasDataI(10,10,f0,f1,0.1)
```

If we want to generate the genotypes of 10 cases and 10 controls with pairwise correlationship 0.1 within cases and 0.2 within controls at 20 SNPs with MAFs generated uniformly from [0.05,0.5], we can:


## Example 6: Condition 3

```{r}
f0=f1<-runif(20,0.05,0.5)
simGwasDataII(10,10,f0,f1,0.1,0.2)
```
